File: erlang/etudes/ch6/dates.erl
-module(dates).
-export([date_parts/1,julian/1]).

date_parts(String) -> [ to_integer(S) || S <- re:split(String,"-",[{return,list}]) ].

to_integer(String) -> element(1, string:to_integer(String)).

julian(String) ->
    [Year, Mo, Day] = date_parts(String),
    count_days(Mo, Day, days_in_month(Year)).

count_days(1, Day, _) -> Day;
count_days(Mo, Day, [Head|Tail]) -> Head + count_days(Mo - 1, Day, Tail).

is_leap_year(Year) -> (Year rem 4 == 0 andalso Year rem 100 /= 0) orelse (Year rem 400 == 0).

days_in_month(Year) ->
    case is_leap_year(Year) of
        true -> [31,29,31,30,31,30,31,31,30,31,30,31];
        false -> [31,28,31,30,31,30,31,31,30,31,30,31]
    end.

File: elixir/etudes/ch06/dates.ex
defmodule Dates do
  def date_parts(string), do: Enum.map(String.split(string, "-"), &String.to_integer/1)

  def julian(string) do
    [year, mo, day] = date_parts(string)
    count_days(mo, day, days_in_month(year))
  end

  defp count_days(1, day, _), do: day
  defp count_days(mo, day, [head|tail]), do: head + count_days(mo - 1, day, tail)

  defp is_leap_year(year), do: (rem(year, 4) == 0 and rem(year, 100) != 0) or (rem(year, 400) == 0)

  defp days_in_month(year) do
    if is_leap_year(year), do: [31,29,31,30,31,30,31,31,30,31,30,31], else: [31,28,31,30,31,30,31,31,30,31,30,31]
  end
end
